File: src/incode_mcp/projects.py
"""Project marker creation and resolution."""

from __future__ import annotations

import tomllib
from collections.abc import Iterable
from pathlib import Path
from uuid import UUID, uuid4

import tomli_w
from pydantic import ValidationError

from .errors import ErrorCode, IncodeError
from .models import DEFAULT_INCLUDES, LEGACY_DEFAULT_INCLUDES_V1, ProjectInfo, ScanConfig

MARKER_DIRECTORY = ".ci-mcp"
LEGACY_MARKER_DIRECTORY = ".incode"
MARKER_FILE = "project.toml"


def marker_path(root: Path) -> Path:
    return root / MARKER_DIRECTORY / MARKER_FILE


def legacy_marker_path(root: Path) -> Path:
    return root / LEGACY_MARKER_DIRECTORY / MARKER_FILE


def existing_marker_path(root: Path) -> Path | None:
    current = marker_path(root)
    if current.is_file():
        return current
    legacy = legacy_marker_path(root)
    if legacy.is_file():
        return legacy
    return None


def initialize_project(
    root: Path, *, name: str | None = None, force_new_id: bool = False
) -> ProjectInfo:
    root = root.expanduser().resolve()
    if not root.is_dir():
        raise IncodeError(ErrorCode.PROJECT_NOT_FOUND, f"Project directory does not exist: {root}")
    if existing_marker_path(root) is not None and not force_new_id:
        return read_project_marker(root)

    existing_path = marker_path(root)
    project = ProjectInfo(id=str(uuid4()), name=name or root.name, root=root)
    directory = existing_path.parent
    directory.mkdir(mode=0o700, parents=True, exist_ok=True)
    (directory / ".gitignore").write_text("*\n", encoding="utf-8")
    data = {
        "version": project.version,
        "id": project.id,
        "name": project.name,
        "scan": project.scan.model_dump(),
    }
    existing_path.write_text(tomli_w.dumps(data), encoding="utf-8")
    return project


def read_project_marker(root: Path) -> ProjectInfo:
    root = root.expanduser().resolve()
    path = existing_marker_path(root) or marker_path(root)
    try:
        raw = tomllib.loads(path.read_text(encoding="utf-8"))
        if raw.get("version") != 1:
            raise ValueError("unsupported marker version")
        UUID(str(raw["id"]))
        scan = ScanConfig.model_validate(raw.get("scan", {}))
        if scan.include == LEGACY_DEFAULT_INCLUDES_V1:
            scan = scan.model_copy(update={"include": list(DEFAULT_INCLUDES)})
        return ProjectInfo(
            version=raw["version"],
            id=str(raw["id"]),
            name=str(raw["name"]),
            root=root,
            scan=scan,
        )
    except (OSError, KeyError, TypeError, ValueError, ValidationError) as exc:
        raise IncodeError(
            ErrorCode.PROJECT_NOT_FOUND,
            f"Invalid or missing project marker: {path}",
            path=str(path),
        ) from exc


def find_project_root(start: Path) -> Path | None:
    current = start.expanduser().resolve()
    if current.is_file():
        current = current.parent
    for candidate in (current, *current.parents):
        if existing_marker_path(candidate) is not None:
            return candidate
    return None


class ProjectResolver:
    def __init__(self, projects: Iterable[ProjectInfo]) -> None:
        self._projects = list(projects)

    def resolve(
        self,
        *,
        explicit: str | None = None,
        roots: Iterable[Path] = (),
        cwd: Path | None = None,
    ) -> ProjectInfo:
        if explicit:
            return self._resolve_explicit(explicit)

        roots = list(roots)
        marked = self._marked_projects(roots)
        if len(marked) == 1:
            return marked[0]
        if len(marked) > 1:
            raise IncodeError(
                ErrorCode.AMBIGUOUS_PROJECT,
                "Multiple MCP roots contain initialized projects",
                projects=[project.id for project in marked],
            )

        if cwd is not None and (root := find_project_root(cwd)) is not None:
            return self._by_root_or_marker(root)
        raise IncodeError(
            ErrorCode.PROJECT_NOT_FOUND,
            "No active Incode project was detected; pass an explicit project id, name, or "
            "path, or run init_project for this directory",
            searched_roots=[str(root) for root in roots],
        )

    def _resolve_explicit(self, explicit: str) -> ProjectInfo:
        direct = [
            project
            for project in self._projects
            if project.id == explicit or project.name == explicit
        ]
        if len(direct) == 1:
            return direct[0]
        if len(direct) > 1:
            raise IncodeError(
                ErrorCode.AMBIGUOUS_PROJECT,
                f"Project name is ambiguous: {explicit}",
                projects=[project.id for project in direct],
            )
        candidate = Path(explicit).expanduser()
        if candidate.exists():
            root = find_project_root(candidate)
            if root is not None:
                return self._by_root_or_marker(root)
        raise IncodeError(ErrorCode.PROJECT_NOT_FOUND, f"Unknown project: {explicit}")

    def _marked_projects(self, roots: Iterable[Path]) -> list[ProjectInfo]:
        found: dict[str, ProjectInfo] = {}
        for candidate in roots:
            root = find_project_root(candidate)
            if root is not None:
                project = self._by_root_or_marker(root)
                found[project.id] = project
        return list(found.values())

    def _by_root_or_marker(self, root: Path) -> ProjectInfo:
        resolved = root.resolve()
        for project in self._projects:
            if project.root.resolve() == resolved:
                return project
        return read_project_marker(resolved)

File: ts/packages/server/src/projects.ts
/** Project marker creation and resolution. */

import { randomUUID } from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { parse as parseToml, stringify as stringifyToml } from "smol-toml";
import { CodeIndexingError } from "./errors.ts";
import {
  DEFAULT_INCLUDES,
  LEGACY_DEFAULT_INCLUDES_V1,
  LEGACY_DEFAULT_INCLUDES_V2,
  LEGACY_DEFAULT_INCLUDES_V3,
  ProjectInfo,
  ScanConfig,
} from "./models.ts";
import {
  expandUser,
  fileIdentity,
  resolvePath,
  rootedUnder,
  sameFile as sameProjectRoot,
} from "./paths.ts";

// Defined in paths.ts, where the pathlib semantics they depend on live, and
// exported from here because this is the module that gives them their meaning.
export { fileIdentity as projectRootIdentity, rootedUnder, sameProjectRoot };

export const MARKER_DIRECTORY = ".ci-mcp";
export const LEGACY_MARKER_DIRECTORY = ".code-indexing-mcp";
export const MARKER_FILE = "project.toml";

export function markerPath(root: string): string {
  return path.join(root, MARKER_DIRECTORY, MARKER_FILE);
}

export function legacyMarkerPath(root: string): string {
  return path.join(root, LEGACY_MARKER_DIRECTORY, MARKER_FILE);
}

function isFile(value: string): boolean {
  try {
    return fs.statSync(value).isFile();
  } catch {
    return false;
  }
}

function isDirectory(value: string): boolean {
  try {
    return fs.statSync(value).isDirectory();
  } catch {
    return false;
  }
}

export function existingMarkerPath(root: string): string | null {
  const current = markerPath(root);
  if (isFile(current)) return current;
  const legacy = legacyMarkerPath(root);
  if (isFile(legacy)) return legacy;
  return null;
}

export function initializeProject(
  root: string,
  { name, forceNewId = false }: { name?: string | undefined; forceNewId?: boolean } = {},
): ProjectInfo {
  const resolved = resolvePath(root);
  if (!isDirectory(resolved)) {
    throw new CodeIndexingError(
      "PROJECT_NOT_FOUND",
      `Project directory does not exist: ${resolved}`,
    );
  }
  if (existingMarkerPath(resolved) !== null && !forceNewId) {
    return readProjectMarker(resolved);
  }

  const target = markerPath(resolved);
  const project = ProjectInfo.parse({
    id: randomUUID(),
    name: name ?? path.basename(resolved),
    root: resolved,
  });
  const directory = path.dirname(target);
  fs.mkdirSync(directory, { recursive: true, mode: 0o700 });
  fs.writeFileSync(path.join(directory, ".gitignore"), "*\n", "utf8");
  fs.writeFileSync(
    target,
    stringifyToml({
      version: project.version,
      id: project.id,
      name: project.name,
      scan: project.scan,
    }),
    "utf8",
  );
  return project;
}

export function readProjectMarker(root: string): ProjectInfo {
  const resolved = resolvePath(root);
  const target = existingMarkerPath(resolved) ?? markerPath(resolved);
  try {
    const raw = parseToml(fs.readFileSync(target, "utf8"));
    if (raw.version !== 1) throw new Error("unsupported marker version");
    if (raw.id === undefined) throw new Error("marker has no id");
    const id = String(raw.id);
    if (!isUuid(id)) throw new Error("marker id is not a uuid");
    if (raw.name === undefined) throw new Error("marker has no name");
    let scan = ScanConfig.parse(raw.scan ?? {});
    // A marker still carrying an older default include list is upgraded to the
    // current one, so a project written before a language was supported picks it
    // up. An include list the user has edited is left exactly as written -- and
    // the file itself is never rewritten either way.
    if (isLegacyDefaultIncludes(scan.include)) {
      scan = { ...scan, include: [...DEFAULT_INCLUDES] };
    }
    return ProjectInfo.parse({
      version: raw.version,
      id,
      name: String(raw.name),
      root: resolved,
      scan,
    });
  } catch (cause) {
    throw new CodeIndexingError(
      "PROJECT_NOT_FOUND",
      `Invalid or missing project marker: ${target}`,
      { path: target },
      { cause },
    );
  }
}

function isLegacyDefaultIncludes(include: readonly string[]): boolean {
  return [LEGACY_DEFAULT_INCLUDES_V1, LEGACY_DEFAULT_INCLUDES_V2, LEGACY_DEFAULT_INCLUDES_V3].some(
    (legacy) =>
      legacy.length === include.length &&
      legacy.every((pattern, index) => include[index] === pattern),
  );
}

/**
 * The spellings Python's `UUID()` accepts, which is what wrote these markers.
 *
 * Wider than the canonical form on purpose: a marker hand-edited to
 * `{...}` or a bare 32-hex-digit id validated under the Python build and must
 * keep validating, or the project would read as uninitialized after an upgrade.
 */
function isUuid(value: string): boolean {
  const stripped = value
    .trim()
    .replace(/^urn:uuid:/i, "")
    .replace(/^\{(.*)\}$/, "$1")
    .replaceAll("-", "");
  return /^[0-9a-f]{32}$/i.test(stripped);
}

export function findProjectRoot(start: string): string | null {
  let current = resolvePath(start);
  if (isFile(current)) current = path.dirname(current);
  for (;;) {
    if (existingMarkerPath(current) !== null) return current;
    const parent = path.dirname(current);
    if (parent === current) return null;
    current = parent;
  }
}

export class ProjectResolver {
  readonly #projects: readonly ProjectInfo[];

  constructor(projects: Iterable<ProjectInfo>) {
    this.#projects = [...projects];
  }

  resolve({
    explicit,
    roots = [],
    cwd,
  }: {
    explicit?: string | undefined;
    roots?: readonly string[];
    cwd?: string | undefined;
  } = {}): ProjectInfo {
    if (explicit) return this.#resolveExplicit(explicit);

    const marked = this.#markedProjects(roots);
    if (marked.length === 1) return marked[0] as ProjectInfo;
    if (marked.length > 1) {
      throw new CodeIndexingError(
        "AMBIGUOUS_PROJECT",
        "Multiple MCP roots contain initialized projects",
        { projects: marked.map((project) => project.id) },
      );
    }

    if (cwd !== undefined) {
      const root = findProjectRoot(cwd);
      if (root !== null) return this.#byRootOrMarker(root);
    }
    throw new CodeIndexingError(
      "PROJECT_NOT_FOUND",
      "No active CodeIndexing project was detected; pass an explicit project id, name, or " +
        "path, or run init_project for this directory",
      { searched_roots: [...roots] },
    );
  }

  #resolveExplicit(explicit: string): ProjectInfo {
    const direct = this.#projects.filter(
      (project) => project.id === explicit || project.name === explicit,
    );
    if (direct.length === 1) return direct[0] as ProjectInfo;
    if (direct.length > 1) {
      throw new CodeIndexingError("AMBIGUOUS_PROJECT", `Project name is ambiguous: ${explicit}`, {
        projects: direct.map((project) => project.id),
      });
    }
    const candidate = expandUser(explicit);
    if (fs.existsSync(candidate)) {
      const root = findProjectRoot(candidate);
      if (root !== null) return this.#byRootOrMarker(root);
    }
    throw new CodeIndexingError("PROJECT_NOT_FOUND", `Unknown project: ${explicit}`);
  }

  #markedProjects(roots: readonly string[]): ProjectInfo[] {
    const found = new Map<string, ProjectInfo>();
    for (const candidate of roots) {
      const root = findProjectRoot(candidate);
      if (root !== null) {
        const project = this.#byRootOrMarker(root);
        found.set(project.id, project);
      }
    }
    return [...found.values()];
  }

  #byRootOrMarker(root: string): ProjectInfo {
    const resolved = resolvePath(root);
    for (const project of this.#projects) {
      if (sameProjectRoot(project.root, resolved)) return project;
    }
    return readProjectMarker(resolved);
  }
}
